osquery/tables/system/windows/scheduled_tasks.cpp

```diff
@@ -0,0 +1,160 @@
+  if (FAILED(ret)) {
+    VLOG(1) << "Failed to get root task folder " << ret;
+    return;
+  }
+
+  IRegisteredTaskCollection* pTaskCollection = nullptr;
+  ret = pRootFolder->GetTasks(TASK_ENUM_HIDDEN, &pTaskCollection);
+  pRootFolder->Release();
+  if (FAILED(ret)) {
+    VLOG(1) << "Failed to get task collection for root folder " << ret;
+    return;
+  }
+
+  long numTasks = 0;
+  pTaskCollection->get_Count(&numTasks);
+  for (size_t i = 0; i < numTasks; i++) {
+    IRegisteredTask* pRegisteredTask = nullptr;
+    // Colletions are 1-based lists
+    ret = pTaskCollection->get_Item(_variant_t(i + 1), &pRegisteredTask);
+    if (FAILED(ret)) {
+      VLOG(1) << "Failed to process task";
+      continue;
+    }
+
+    Row r;
+    BSTR taskName;
+    ret = pRegisteredTask->get_Name(&taskName);
+    std::wstring wTaskName(taskName, SysStringLen(taskName));
+    r["name"] = ret == S_OK ? SQL_TEXT(wstringToString(wTaskName.c_str())) : "";
+
+    VARIANT_BOOL enabled = false;
+    pRegisteredTask->get_Enabled(&enabled);
+    r["enabled"] = enabled ? INTEGER(1) : INTEGER(0);
+
+    TASK_STATE taskState;
+    pRegisteredTask->get_State(&taskState);
+    r["state"] = kStateMap.count(taskState) > 0
+                     ? kStateMap.at(taskState)
+                     : kStateMap.at(TASK_STATE_UNKNOWN);
+
+    BSTR taskPath;
+    ret = pRegisteredTask->get_Path(&taskPath);
+    std::wstring wTaskPath(taskPath, SysStringLen(taskPath));
+    r["path"] = ret == S_OK ? SQL_TEXT(wstringToString(wTaskPath.c_str())) : "";
+
+    VARIANT_BOOL hidden = false;
+    pRegisteredTask->get_Enabled(&hidden);
+    r["hidden"] = hidden ? INTEGER(1) : INTEGER(0);
+
+    HRESULT lastTaskRun = E_FAIL;
+    pRegisteredTask->get_LastTaskResult(&lastTaskRun);
+    _com_error err(lastTaskRun);
+    r["last_run_message"] = err.ErrorMessage();
+    r["last_run_code"] = INTEGER(lastTaskRun);
+
+    DATE dRunTime;
+    pRegisteredTask->get_LastRunTime(&dRunTime);
+    COleDateTime cdt(dRunTime);
+    r["last_run_time"] = cdt.Format(kTimeFormat.c_str());
+
+    pRegisteredTask->get_NextRunTime(&dRunTime);
+    cdt = dRunTime;
+    r["next_run_time"] = cdt.Format(kTimeFormat.c_str());
+
+    results.push_back(r);
+    pRegisteredTask->Release();
+  }
+  pTaskCollection->Release();
+}
+
+QueryData genScheduledTasks(QueryContext& context) {
+  QueryData results;
+
+  // First process all tasks in the root folder
+  enumerateTasksForFolder("\\", results);
+
+  // We attempt to derive the location of the tasks folder
+  auto sysRoot = getEnvVar("SystemRoot");
+  if (!sysRoot.is_initialized()) {
+    return results;
+  }
+  auto sysTaskPath = *sysRoot + "\\System32\\Tasks";
+
+  // Then process all tasks in subsequent folders
+  std::vector<std::string> taskPaths;
+  listDirectoriesInDirectory(sysTaskPath, taskPaths, true);
+  for (const auto& path : taskPaths) {
+    auto taskPath =
+        "\\" +
+        join(split(path.substr(sysTaskPath.size(), path.size()), "\\"), "\\");
```

Comment:
While nothing pops out immediately, it'd be good to double check this logic, is there anything that could cause `.substr` to throw an exception?